java/src/main/java/org/codingdojo/yatzy1/Roll.java

```diff
@@ -17,22 +17,35 @@ public IntStream filterByValue(int dieValue) {
             .mapToInt(Integer::intValue);
     }
 
-    public Map<Integer, Integer> counts() {
-        Map<Integer, Integer> map = new HashMap<>();
-        for (int i = 1; i <= 6; i++) {
-            map.put(i, 0);
+    public List<Integer> filterByCount(IntPredicate countPredicate) {
+        var counts = counts();
+        var result = new ArrayList<Integer>();
+        for (int i = 0; i < counts.length; i++) {
+            int dieValue = i + 1;
+            if(countPredicate.test(counts[i])) result.add(dieValue);
         }
-        for (int die : dice) {
-            map.put(die, map.get(die) + 1);
-        }
-        return map;
+        return result;
     }
 
-    public Set<Integer> diceOfAKind(int numberOfDiceWithSameValue) {
-        return counts().entrySet().stream()
-            .filter(entry -> entry.getValue() >= numberOfDiceWithSameValue)
-            .map(Map.Entry::getKey)
-            .collect(Collectors.toSet());
+    /*
+    In terms of complexity, both map and array are equivalent,
+    with O(n) time complexity and O(1) space complexity for the additional structures they use.
+    However, the array-based approach (int[]) is likely to be more efficient in practice for several reasons:
+        - Performance:
+          Array access and updates are generally faster than HashMap operations
+          because arrays have lower overhead and better cache locality.
```

Comment:
"using int[] can be more efficient due to the lack of autoboxing/unboxing overhead, making it preferable for performance-critical applications dealing with large datasets. "